File: apps/app-frontend/src/helpers/curseforge.ts
/**
 * CurseForge API integration helpers.
 *
 * Wraps the `plugin:curseforge|*` Tauri commands exposed by the Rust backend
 * in `apps/app/src/api/curseforge.rs`. CurseForge search results are converted
 * to a Modrinth V3-compatible shape on the backend, so the rest of the UI can
 * render them like normal Modrinth results.
 */

import { invoke } from '@tauri-apps/api/core'

import type { InstallJobSnapshot } from './install'

/** A single CurseForge project (mod/modpack/resourcepack/shader/world/datapack). */
export interface CfMod {
	id: number
	game_id: number
	name: string
	slug: string
	summary: string
	download_count: number
	primary_category_id: number
	class_id: number | null
	categories: CfCategory[]
	authors: CfAuthor[]
	logo: CfAsset | null
	screenshots: CfAsset[]
	links: CfLinks
	date_created: string
	date_modified: string
	date_released: string
	allow_mod_distribution: boolean | null
	game_popularity_rank: number | null
	is_available: boolean | null
	latest_files: CfFile[]
	latest_files_indexes: CfFileIndex[]
}

export interface CfCategory {
	id: number
	name: string
	slug: string
	class_id: number | null
	parent_category_id: number | null
	icon_url: string | null
}

export interface CfAuthor {
	id: number
	name: string
	url: string | null
}

export interface CfAsset {
	id: number
	url: string | null
	thumbnail_url: string | null
}

export interface CfLinks {
	website_url: string | null
	wiki_url: string | null
	issues_url: string | null
	source_url: string | null
}

export interface CfFile {
	id: number
	mod_id: number
	is_available: boolean | null
	display_name: string
	file_name: string
	release_type: number // 1=Release, 2=Beta, 3=Alpha
	file_status: number
	file_date: string
	file_length: number
	download_url: string | null
	game_versions: string[]
	sortable_game_versions: CfSortableGameVersion[]
	dependencies: CfFileDependency[]
	alternate_file_id: number
	is_server_pack: boolean | null
	file_fingerprint: number
}

export interface CfFileIndex {
	game_version: string
	file_id: number | null
	filename: string | null
	release_type: number | null
	game_version_type_id: number | null
	mod_loader: number | null
}

export interface CfSortableGameVersion {
	game_version_name: string
	game_version_padded: string
	game_version: string
	game_version_release_date: string | null
	game_version_type_id: number | null
}

export interface CfFileDependency {
	mod_id: number
	file_id: number | null
	relation_type: number
}

/**
 * Modrinth V3-compatible search result hit returned by `cf_search`.
 *
 * CurseForge-specific extras are prefixed with `_cf_` so frontend code can
 * detect the source when needed while reusing all the regular V3 fields.
 */
export interface CfSearchHit {
	project_id: string
	version_id: string
	project_types: string[]
	slug: string
	author: string
	author_id: string
	organization: string | null
	organization_id: string | null
	name: string
	summary: string
	categories: string[]
	display_categories: string[]
	downloads: number
	follows: number
	icon_url: string | null
	date_created: string
	date_modified: string
	license: string
	gallery: string[]
	featured_gallery: string | null
	color: number | null
	loaders: string[]
	versions: string[]
	// CurseForge-specific extras
	_cf_mod_id: number
	_cf_class_id: number
	_cf_page_url: string | null
	_cf_download_count: number
}

export interface CfSearchResults {
	search: string
	result: {
		hits: CfSearchHit[]
		offset: number
		limit: number
		total_hits: number
	}
}

export interface CfSearchParams {
	/** "mod" | "modpack" | "resourcepack" | "shader" | "world" | "datapack" */
	projectType: string
	searchFilter?: string | null
	gameVersion?: string | null
	/** "forge" | "fabric" | "quilt" | "neoforge" | "liteloader" */
	loader?: string | null
	/** "relevance" | "downloads" | "follows" | "newest" | "updated" */
	sort?: string | null
	/** 0-indexed page */
	page?: number | null
	pageSize?: number | null
}

/** Search CurseForge; returns Modrinth V3-compatible results. */
export async function cf_search(params: CfSearchParams): Promise<CfSearchResults> {
	return await invoke<CfSearchResults>('plugin:curseforge|cf_search', { params })
}

/** Get a single CurseForge project by mod ID. */
export async function cf_get_mod(modId: number): Promise<CfMod> {
	return await invoke<CfMod>('plugin:curseforge|cf_get_mod', { modId })
}

/** Get files for a CurseForge project, optionally filtered by game version and loader. */
export async function cf_get_mod_files(
	modId: number,
	gameVersion?: string | null,
	loader?: string | null,
): Promise<CfFile[]> {
	return await invoke<CfFile[]>('plugin:curseforge|cf_get_mod_files', {
		modId,
		gameVersion: gameVersion ?? null,
		loader: loader ?? null,
	})
}

/** Get a single CurseForge file by mod ID and file ID. */
export async function cf_get_file(modId: number, fileId: number): Promise<CfFile> {
	return await invoke<CfFile>('plugin:curseforge|cf_get_file', { modId, fileId })
}

/** Get the direct download URL for a CurseForge file. Returns null if unavailable. */
export async function cf_get_file_download_url(
	modId: number,
	fileId: number,
): Promise<string | null> {
	return await invoke<string | null>('plugin:curseforge|cf_get_file_download_url', {
		modId,
		fileId,
	})
}

/** Get CurseForge categories. If `projectType` is null, returns all categories for Minecraft. */
export async function cf_get_categories(
	projectType?: string | null,
): Promise<CfCategory[]> {
	return await invoke<CfCategory[]>('plugin:curseforge|cf_get_categories', {
		projectType: projectType ?? null,
	})
}

/**
 * Install a CurseForge file to an existing instance via the InstallJob system.
 * The download popup shows progress like other installs.
 *
 * @param contentType "mod" | "resourcepack" | "shader" | "datapack" | "world" | "modpack"
 */
export async function cf_install_file(
	instanceId: string,
	modId: number,
	fileId: number,
	fileName: string,
	downloadUrl: string | null,
	contentType: string,
	title: string,
	iconUrl: string | null,
): Promise<InstallJobSnapshot> {
	return await invoke<InstallJobSnapshot>('plugin:curseforge|cf_install_file', {
		instanceId,
		modId,
		fileId,
		fileName,
		downloadUrl,
		contentType,
		title,
		iconUrl,
	})
}

File: apps/app/src/api/curseforge.rs
//! Tauri plugin bindings for CurseForge API integration.
//!
//! Exposes curseforge search, project lookup, file listing, and a
//! InstallJob-driven file installer to the frontend.

use crate::api::Result;
use serde::Deserialize;
use theseus::curseforge::{
    CfClassId, CfFile, CfMod, CfModLoaderType, CfSortField, CfSortOrder,
    CfSearchParams,
};
use theseus::install::InstallJobSnapshot;
use theseus::prelude::*;

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("curseforge")
        .invoke_handler(tauri::generate_handler![
            cf_search,
            cf_get_mod,
            cf_get_mod_files,
            cf_get_file,
            cf_get_file_download_url,
            cf_get_categories,
            cf_install_file,
        ])
        .build()
}

/// Search parameters mirroring `CfSearchParams`, with camelCase fields
/// so the frontend can pass plain JS objects.
#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CfSearchParamsRequest {
    /// Project type string ("mod"|"modpack"|"resourcepack"|"shader"|"world"|"datapack")
    pub project_type: String,
    #[serde(default)]
    pub search_filter: Option<String>,
    #[serde(default)]
    pub game_version: Option<String>,
    /// Loader name ("forge"|"fabric"|"quilt"|"neoforge"|"liteloader") or null
    #[serde(default)]
    pub loader: Option<String>,
    /// Modrinth sort name ("relevance"|"downloads"|"follows"|"newest"|"updated")
    #[serde(default)]
    pub sort: Option<String>,
    /// 0-indexed page
    #[serde(default)]
    pub page: Option<u32>,
    #[serde(default)]
    pub page_size: Option<u32>,
}

/// Search CurseForge and return Modrinth V3-compatible results.
#[tauri::command]
pub async fn cf_search(params: CfSearchParamsRequest) -> Result<SearchResultsV3> {
    let state = State::get().await?;
    let class_id = CfClassId::from_project_type(&params.project_type)
        .ok_or_else(|| {
            theseus::Error::from(theseus::ErrorKind::InputError(format!(
                "Unknown project_type: {}",
                params.project_type
            )))
        })?;

    let mod_loader_type = params
        .loader
        .as_deref()
        .and_then(CfModLoaderType::from_loader_name);

    let sort_field = params
        .sort
        .as_deref()
        .map(CfSortField::from_modrinth_sort);

    let search_params = CfSearchParams {
        class_id,
        search_filter: params.search_filter,
        game_version: params.game_version,
        mod_loader_type,
        sort_field,
        sort_order: Some(CfSortOrder::Descending),
        page: params.page,
        page_size: params.page_size,
    };

    Ok(theseus::curseforge::search(search_params, &state).await?)
}

/// Get a single CurseForge mod by ID.
#[tauri::command]
pub async fn cf_get_mod(mod_id: i64) -> Result<CfMod> {
    let state = State::get().await?;
    Ok(theseus::curseforge::get_mod(mod_id, &state).await?)
}

/// Get files for a CurseForge mod, optionally filtered.
#[tauri::command]
pub async fn cf_get_mod_files(
    mod_id: i64,
    game_version: Option<String>,
    loader: Option<String>,
) -> Result<Vec<CfFile>> {
    let state = State::get().await?;
    let loader_type = loader
        .as_deref()
        .and_then(CfModLoaderType::from_loader_name);
    Ok(theseus::curseforge::get_mod_files(
        mod_id,
        game_version.as_deref(),
        loader_type,
        &state,
    )
    .await?)
}

/// Get a single CurseForge file by mod_id and file_id.
#[tauri::command]
pub async fn cf_get_file(mod_id: i64, file_id: i64) -> Result<CfFile> {
    let state = State::get().await?;
    Ok(theseus::curseforge::get_file(mod_id, file_id, &state).await?)
}

/// Get the direct download URL for a CurseForge file.
/// Returns null if the file is not available for third-party download.
#[tauri::command]
pub async fn cf_get_file_download_url(
    mod_id: i64,
    file_id: i64,
) -> Result<Option<String>> {
    let state = State::get().await?;
    Ok(theseus::curseforge::get_file_download_url(mod_id, file_id, &state)
        .await?)
}

/// Get CurseForge categories for a given class.
/// If project_type is null, returns all categories for Minecraft.
#[tauri::command]
pub async fn cf_get_categories(
    project_type: Option<String>,
) -> Result<Vec<theseus::curseforge::CfCategory>> {
    let state = State::get().await?;
    let class_id = project_type
        .as_deref()
        .and_then(CfClassId::from_project_type);
    Ok(theseus::curseforge::get_categories(class_id, &state).await?)
}

/// Install a CurseForge file to an existing instance via the InstallJob
/// system. Displays progress in the download popup like other installs.
#[tauri::command]
pub async fn cf_install_file(
    instance_id: String,
    mod_id: i64,
    file_id: i64,
    file_name: String,
    download_url: Option<String>,
    content_type: String,
    title: String,
    icon_url: Option<String>,
) -> Result<InstallJobSnapshot> {
    Ok(theseus::install::install_curseforge_file(
        instance_id,
        mod_id,
        file_id,
        file_name,
        download_url,
        content_type,
        title,
        icon_url,
    )
    .await?)
}
